queue/driver_mongo.go

```diff
@@ -1400,14 +1268,146 @@ type dispatchAttemptInfo struct {
 	// misses is the number of times the dispatcher attempted to dispatch a job
 	// but failed.
 	misses int
+	// total is the total number of documents the cursor returned.
+	total int
+}
+
+// dispatchStaleAndPending attempts to dispatch a job. Stale in progress jobs are given precedence over pending jobs.
+func (d *mongoDriver) dispatchStaleAndPending(ctx context.Context, startAt time.Time) (amboy.Job, dispatchAttemptInfo, error) {
+	now := time.Now()
+	var info dispatchAttemptInfo
+
+	job, dispatchInfo, err := d.tryDispatchJob(ctx, d.getNextStaleInProgressQuery(now), d.opts.SampleSize, startAt)
+	info.misses += dispatchInfo.misses
+	info.skips += dispatchInfo.skips
+	if err != nil {
+		return nil, info, errors.Wrap(err, "dispatching stale in progress job")
+	}
+
+	if job != nil || d.opts.SampleSize > 0 && dispatchInfo.total >= d.opts.SampleSize {
+		return job, info, nil
+	}
+
+	pendingSampleSize := 0
+	if d.opts.SampleSize > 0 {
```

Comment:
I also think it's okay to not reduce the sample size for the follow-up query. The main purpose of the field is to cap the number of jobs checked per query, not cap the total number of jobs checked per `Next` round. We can just change the doc comment on `SampleSize` to indicate that it's the number of jobs per set of jobs checked rather than total.

Reply:
done.
That made it a lot easier!